internal/adapters/storage/redis/room/redis.go

```diff
@@ -440,3 +481,56 @@ func getRoomStatusUpdateChannel(scheduler, roomID string) string {
 func getRoomOccupancyRedisKey(scheduler string) string {
 	return fmt.Sprintf("scheduler:%s:occupancy", scheduler)
 }
+
+func (r *redisStateStorage) AllocateRoom(ctx context.Context, schedulerName string) (string, error) {
+	statusKey := getRoomStatusSetRedisKey(schedulerName)
+	roomKeyPrefix := fmt.Sprintf("scheduler:%s:rooms:", schedulerName)
+	readyStatus := int(game_room.GameStatusReady)
+	allocatedStatus := int(game_room.GameStatusAllocated)
+	currentTime := time.Now().Unix()
+
+	var result interface{}
+	var err error
+	metrics.RunWithMetrics(roomStorageMetricLabel, func() error {
+		result, err = allocateRoomScript.Run(ctx, r.client, []string{statusKey, roomKeyPrefix}, readyStatus, allocatedStatus, currentTime).Result()
+		return err
+	})
+
+	if err != nil {
+		if err == redis.Nil {
+			return "", nil
+		}
+		return "", errors.NewErrUnexpected("error executing room allocation script for scheduler %s", schedulerName).WithError(err)
+	}
+
+	if result == nil {
+		return "", nil
+	}
+
+	roomID, ok := result.(string)
+	if !ok {
+		return "", errors.NewErrUnexpected("unexpected result type from room allocation script for scheduler %s", schedulerName)
+	}
+
+	if roomID != "" {
+		encodedEvent, err := encodeStatusEvent(&game_room.StatusEvent{
+			RoomID:        roomID,
+			SchedulerName: schedulerName,
+			Status:        game_room.GameStatusAllocated,
+		})
+		if err != nil {
+			return "", errors.NewErrEncoding("failed to encode status event for allocated room").WithError(err)
+		}
+
+		var publishCmd *redis.IntCmd
+		metrics.RunWithMetrics(roomStorageMetricLabel, func() error {
+			publishCmd = r.client.Publish(ctx, getRoomStatusUpdateChannel(schedulerName, roomID), encodedEvent)
+			return publishCmd.Err()
+		})
+		if publishCmd.Err() != nil {
+			return "", errors.NewErrUnexpected("error sending room allocation status event for room %s", roomID).WithError(publishCmd.Err())
+		}
+	}
```

Comment:
I'd move this out to another function, it will simplify to caller to treat these errors individually